flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,180 @@
+			final ConnectionID connectionId2 = new ConnectionID(new InetSocketAddress("localhost", 5000), 0);
+
+			RemoteInputChannel remoteInputChannel1 = createUnknownInputChannel(network, inputGate1, resultPartitionId1, 0).toRemoteInputChannel(connectionId1);
+			inputGate1.setInputChannel(resultPartitionId1.getPartitionId(), remoteInputChannel1);
+
+			RemoteInputChannel remoteInputChannel2 = createUnknownInputChannel(network, inputGate1, resultPartitionId2, 1).toRemoteInputChannel(connectionId2);
+			inputGate1.setInputChannel(resultPartitionId2.getPartitionId(), remoteInputChannel2);
+
+			RemoteInputChannel remoteInputChannel3 = createUnknownInputChannel(network, inputGate2, resultPartitionId1, 0).toRemoteInputChannel(connectionId1);
+			inputGate2.setInputChannel(resultPartitionId1.getPartitionId(), remoteInputChannel3);
+
+			LocalInputChannel localInputChannel1 = createUnknownInputChannel(network, inputGate2, resultPartitionId2, 1).toLocalInputChannel();
+			inputGate2.setInputChannel(resultPartitionId2.getPartitionId(), localInputChannel1);
+
+			MetricGroup inputMetric = UnregisteredMetricGroups.createUnregisteredTaskMetricGroup().getIOMetricGroup().addGroup("buffers");
+
+			inputGate1.setup();
+			inputGate2.setup();
+
+			SingleInputGate[] inputGates = new SingleInputGate[]{inputGate1, inputGate2};
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBufferPoolUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBufferPoolUsageGauge(floatingBuffersUsageGauge, exclusiveBuffersUsageGauge, inputGates);
+
+			inputMetric.gauge("floatingBuffersUsage", floatingBuffersUsageGauge);
+			inputMetric.gauge("exclusiveBuffersUsage", exclusiveBuffersUsageGauge);
+			inputMetric.gauge("inputBufferUsage", inputBufferPoolUsageGauge);
+
+			assertEquals(0.0, floatingBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, exclusiveBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBufferPoolUsageGauge.getValue(), 0.0);
+
+			assertEquals(buffersPerChannel * 2, exclusiveBuffersUsageGauge.calculateBufferPoolSize(inputGate1));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateBufferPoolSize(inputGate1));
+
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateBufferPoolSize(inputGate2));
+			assertEquals(extraNetworkBuffersPerGate, floatingBuffersUsageGauge.calculateBufferPoolSize(inputGate2));
+
+			// test for exclusive buffer used
+			drainBuffer(buffersPerChannel, remoteInputChannel1, closeableRegistry);
+
+			assertEquals(buffersPerChannel, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate1));
+			assertEquals(1 / 3.0, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+			assertEquals(buffersPerChannel / (extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0), inputBufferPoolUsageGauge.getValue(), 0.0001);
```

Comment:
We could define a local var for `extraNetworkBuffersPerGate * 2 + buffersPerChannel * 3.0` because it is used many times below.

Reply:
Even we could extract a method for covering the `drainBuffer` with three asserts together if possible, because we do the drain logic three times which might reuse the same logic.